R/function-isotope.R

```diff
@@ -0,0 +1,149 @@
+#' peaks
+#'
+#' @param seed_mz numeric ordered vector containing m/z values. If provided,
+#' the function checks if there are peaks in x whose m/z match them. If so,
+#' it looks for isotope groups on this subsets of peaks.
+#'
+#' @param duplicates character(1), how to handle duplicated matches. Has to be
+#' one of c("keep", "closest", "remove"). No abbreviations allowed.
+#'
+#' @return list of vectors. Each vector in the returned list contains the
+#' indexes of the rows in `x` that match a isotope.
+#'
+#' @details The function iterates over the rows of `x`. For the `i`-th m/z
+#' value in `x` it computes m/z of would-be isotope peaks by adding the `mzd`s
+#' in `isotopeDefinition`. Possible matches of those with the other m/z values
+#' in `x` are then searched. If any is found and the ratio between the
+#' intensities of the corresponding peaks and the `i`-th peak is smaller than a
+#' threshold (specified in the `maxint` column of
+#' `isotopeDefinition`), their indexes are grouped together.
+#'
+#' It can happen that some index is present in more than one group. The user can
+#' decide what to do in this case by choosing the parameter `duplicates`.
+#' `duplicates = "keep"` leaves the groups as they are.
+#' `duplicates = "closest"` removes from the list groups for which another group
+#'  which shares some indexes with it exists and constitutes a better match.
+#' `duplicates = "remove"` removes from the list groups for which another group
+#'  which shares some indexes with it exists
+#'
+#' @examples
+#'
+.isotope_peaks <- function(x, isotopeDefinition = isotopeDefinition(), ppm = 20,
+                           seed_mz = NULL, duplicates = "keep") {
+
+  if(duplicates != "keep" && duplicates != "closest" && duplicates != "remove")
+    stop("'duplicates' has to be one of \"keep\", \"closest\" or \"remove\".")
+
+  lst <- list()
+  idxs_o <- which(x[, 2] > 0)
+  xred <- x[idxs_o, ]
+  if (is.null(seed_mz)) idxs <- 1:nrow(xred)
+  else idxs <- na.omit(MsCoreUtils::closest(seed_mz, xred[, 1], tolerance = 0,
+                                            ppm = ppm, duplicates = "closest"))
+  sgd <- NULL
+  for (i in idxs)
+  {
+    append_to_current <- FALSE
+    for (j in 1:nrow(isotopeDefinition))
+    {
+      tmp <- xred[i, 1] + isotopeDefinition[j, 1]
+      cls <- MsCoreUtils::closest(tmp, xred[-(1:i), 1], tolerance = 0, ppm = ppm)
+      if(!is.na(cls) &  xred[, 2][cls + i] < isotopeDefinition[j, 2]*xred[i, 2])
+      {
+        len <- length(lst)
+        if(!append_to_current)
+        {
+          lst[[len + 1]] <- c(idxs_o[i], idxs_o[cls + i])
+          sgd[len + 1] <- tmp - xred[cls + i, 1]
+          append_to_current <- TRUE
+        }
+        else
+        {
+          lst[[len]] <- c(lst[[len]], idxs_o[cls + i])
+          sgd[len] <- sgd[len] + tmp - xred[cls + i, 1]
+        }
+      }
+    }
+  }
+
+  if(length(lst) > 1 && duplicates == "closest")
```

Comment:
I was thinking if we need this at all or if we should simply avoid duplicates from the start. I think if we start with the smallest m/z and find peaks that match isotope conditions for one peak, these peaks (that were assigned to an isotope cluster) should not be used as "M" peak to search for possible matching isotope peaks. In other words, assume peak with index 4 is our M and peak 6 and 7 are matching the isotope rules, no isotope peaks should be searched for them and they should also not be considered as potential isotope peaks anymore.

What I would suggest is the following alternative approach:

```r
to_test <- rep(TRUE, nrow(x))
full_index <- seq_len(nrow(x))
while(!is.na(i <- match(TRUE, to_test))) {
    tmp <- x[i, 1] + isotopeDefinitions[, 1]
    idxs <- closest(tmp, x[to_test, 1]) # test only against peaks that are not yet in an isotope cluster
    ## BE AWARE THAT idxs IS NOW RELATIVE TO to_test. convert to "real" indices
    if (length(idxs)) {
        idxs <- full_index[to_test][idxs]
        ## check the intensity condition whether they fit
        ints <- x[idxs, 2]
        ...
        ## Disable testing of the M peak as well as of possible matching isotope peaks
        to_test[c(i, idxs[matched])] <- FALSE
    }
}
```

Maybe you find a better way than my approach above, as I guess that this continuous subsetting with `to_test` might have a negative effect on the performance.

Could be that that way we *might* loose some potential isotope peaks, but I would accept that because I think that should be a rare event. It will be however important to clearly explain this in the documentation.

Reply:
Yes, I can do this. Just to see it in an example with the example peak matrix you provided me and the simple two rows isotope definition we lose around 30 groups (over 133 that we obtain with keep parameter) if I'm not mistaken.

Reply:
Uh, that's a lot. But this means that the 30 groups are all overlapping other groups, right? Maybe keep also your original function. I would like to see if we gain anything with this, i.e. how the performance of the two functions are on different examples.

Reply:
There is a significant number of overlappings in the list of groups obtained with "keep" parameter. For example, when I use the parameter "closest" on the same example around 80 groups are removed. Yes, I’ll implement it so that maybe we can compare the results :)